lib/ClangImporter/ImportDecl.cpp

```diff
@@ -2259,6 +2217,44 @@ namespace {
         }
       }
 
+      // We have to do this after populating ImportedDecls to avoid importing
+      // the same decl multiple times. Also after we imported the bases.
+      if (const auto *ctsd =
+              dyn_cast<clang::ClassTemplateSpecializationDecl>(decl)) {
+        for (auto arg : ctsd->getTemplateArgs().asArray()) {
+          llvm::SmallVector<clang::TemplateArgument, 1> nonPackArgs;
+          if (arg.getKind() == clang::TemplateArgument::Pack) {
+            auto pack = arg.getPackAsArray();
+            nonPackArgs.assign(pack.begin(), pack.end());
+          } else {
+            nonPackArgs.push_back(arg);
+          }
+          for (auto realArg : nonPackArgs) {
+            if (realArg.getKind() != clang::TemplateArgument::Type)
+              continue;
+            auto SwiftType = Impl.importTypeIgnoreIUO(
```

Comment:
This code is moved for the sake of consistency. We import base classes above. In case a base class is also mentioned as a template type argument, I think the stack traces are easier to understand if it was imported above as opposed to actually do the first import here. Now we should get a cache hit here. This was not the source of the circular reference.